lib/core/src/Cardano/Wallet.hs

```diff
@@ -1021,23 +1024,71 @@ signCert
     -> Passphrase "encryption"
     -> CoinSelection
     -> PoolId
-    -> ExceptT ErrSignCert IO (Tx, TxMeta, UTCTime, SealedTx)
-signCert _ctx _wid _argGenChange _pwd _coinSel _poolId =
-    error "signCert : unimplemented"
+    -> ExceptT ErrSignDelegationCert IO (Tx, TxMeta, UTCTime, SealedTx)
+signDelegationCert ctx wid argGenChange pwd coinSel _poolId = db & \DBLayer{..} -> do
+    let (CoinSelection ins outs chgs) = coinSel
+    withRootKey @_ @s ctx wid pwd ErrSignDelegationCertWithRootKey $ \xprv -> do
+        mapExceptT atomically $ do
+            cp <- withExceptT ErrSignDelegationCertNoSuchWallet $ withNoSuchWallet wid $
+                readCheckpoint (PrimaryKey wid)
+            let (changeOuts, s') = flip runState (getState cp) $
+                    forM chgs $ \c -> do
+                        addr <- state (genChange argGenChange)
+                        return $ TxOut addr c
+            let allOuts = outs ++ changeOuts
+
+            --deriving chimeric account
+            let accPrv = deriveAccountPrivateKey @k pwd xprv minBound
+            let _addrPrv =
+                    deriveAddressPrivateKey pwd accPrv MutableAccount minBound
+
+            {--
+            let keyFrom = isOwned (getState cp) (xprv, pwd)
+            (tx, wit) <- either
+                (throwE . ErrSignDelegationCert)
+                pure
+                (mkDelegationCertTx tl poolId (addrPrv, pwd) keyFrom ins allOuts)
+            --}
+
+            withExceptT ErrSignDelegationCertNoSuchWallet $
+                putCheckpoint (PrimaryKey wid) (updateState s' cp)
+            let ourCoins (TxOut addr (Coin val)) =
+                    if fst (isOurs addr s')
+                        then Just (fromIntegral val)
+                        else Nothing
+            let amtOuts = sum (mapMaybe ourCoins allOuts)
+            let amtInps = fromIntegral $
+                    sum (getCoin . coin . snd <$> ins)
+            let txSlot =
+                    (currentTip cp) ^. #slotId
+            let meta = TxMeta
+                    { status = Pending
+                    , direction = Outgoing
+                    , slotId = txSlot
+                    , blockHeight = (currentTip cp) ^. #blockHeight
+                    , amount = Quantity (amtInps - amtOuts)
+                    }
+            let time = slotStartTime
+                    (slotParams (blockchainParameters cp))
+                    txSlot
+            return (undefined, meta, time, undefined)
```

Comment:
**[question]**

What about trying to factor the common parts of this and `signTx` in one or more composable functions? I see arguably at least 2 functions:

- change creation

```hs
            let (changeOuts, s') = flip runState (getState cp) $
                    forM chgs $ \c -> do
                        addr <- state (genChange argGenChange)
                        return $ TxOut addr c
            let allOuts = outs ++ changeOuts
```

- metadata construction

```hs
            let ourCoins (TxOut addr (Coin val)) =
                    if fst (isOurs addr s')
                        then Just (fromIntegral val)
                        else Nothing
            let amtOuts = sum (mapMaybe ourCoins allOuts)
            let amtInps = fromIntegral $
                    sum (getCoin . coin . snd <$> ins)
            let txSlot =
                    (currentTip cp) ^. #slotId
            let meta = TxMeta
                    { status = Pending
                    , direction = Outgoing
                    , slotId = txSlot
                    , blockHeight = (currentTip cp) ^. #blockHeight
                    , amount = Quantity (amtInps - amtOuts)
                    }
            let time = slotStartTime
                    (slotParams (blockchainParameters cp))
                    txSlot
```

Perhaps making them separate function internal to the module would help to identify further commonalities?

Reply:
done.